### DISCUSSION TITLE
docs/Libraries-and-Framework

### DISCUSSION THREAD

**Discussion by giscus** [samuelgfeller/blog#3] [Comments]
# docs/Libraries-and-Framework

Libraries and a Framework are essential for building robust and feature-rich applications. But too many dependencies can have a negative impact on the project.

https://samuel-gfeller.ch/docs/Libraries-and-Framework

**Comment by Alpine418** [comment]
Can you please explain why you chose the less used CakePHP database instead of Doctrine's DBAL?

**Reply by samuelgfeller** [reply]
Hi @jnessier, excellent question!

The reason is that the cakephp query builder was the one being used in Odan's [slim4-skeleton](https://github.com/odan/slim4-skeleton) when I started out with this project. I was satisfied with it being so lightweight and straightforward so I had no incentive to change it and thus didn't really explore alternatives. 

I am very open to try out DBAL though and might change it as the default for the [slim-starter](https://samuel-gfeller.ch/slim-starter) and [slim-api-starter](https://samuel-gfeller.ch/slim-api-starter) in the near future.

Would you be interested in a documentation on how to setup and getting doctrine/dbal running with a slim4 application and [dependency injection](https://samuel-gfeller.ch/docs/Dependency-Injection)?

**Reply by Alpine418** [reply]
Hi, thanks for your reply!

I love the Slim Framework and Odan is one of my PHP role models as a hobby developer. I'm currently working on a lightweight ORM library and was looking for a good (and especially secure) replacement for my custom QueryBuilder solution. After studying your Slim Starter project and seeing that Odan also uses QueryBuilder of CakePHP, I decided to go with this library a few days ago. Also, the collection library of CakePHP looks very promising and I'm planning to integrate it into my ORM library as well.

Unfortunately, I'm only a passionate PHP developer in my spare time, usually late at night when the rest of the family is asleep. So my coding time is quite limited and I prefer to use for my ORM and personal web app project. I hope you understand. Still, it makes me proud that you reached out to me as hobby developer.

I wish you a lot of success and hope we can stay in touch - even though I personally see things a bit differently when it comes to your strict interpretation of SOLID.

PS: You can find me on Reddit as Alpine418, where I've already commented on some of your posts in the r/PHP and I'll be using the same username on GitHub soon too. Plus I'm from Switzerland too... so: Hopp Schwiiz :)

**Reply by samuelgfeller** [reply]
Hoi Alpine418 :)

You might be interested in the [conversations](https://samuel-gfeller.ch/docs/Sources-of-Knowledge) I had with Odan over the years! Especially the ones when I kind of started to make my own lightweight ORM [issue #32](https://github.com/odan/slim4-tutorial/issues/32) and [issue #33](https://github.com/odan/slim4-tutorial/issues/33). I also recommend this video from Daniel [Beyond ORM - A guide for advanced database queries](https://www.youtube.com/watch?v=mz6DDGC-SxU). 

If you happen to ask and get an answer from Odan on why he seems to prefer the CakePHP query builder over Doctrine DBAL, please add his response to this thread. 

I wish you a lot of success too and more time for your private projects with PHP!   
I initially had a pretty strong resistance to the ultra strict interpretation of SRP and SOLID too but as I made the example project and talking with Odan with his experience and knowledge I got my eyes opened on how easy and fun programming can be when every feature is decoupled. Code that is more like lego bricks with every path being on its own instead of sticky spider web that is entangled with other things "to simplify initial development". Can totally see how it may seem to make sense (I thought it too) but as the app grows in complexity I felt it in my core how bad of an idea this is. I'm so thankful to Daniel that I didn't have to go through too much hustle to learn this on my own.

Cheers!
